TASK: Deterministic projection schema with sensible defaults

kitsunet opened the pull request:
Projections should be created with deterministic charset and collation and columns containing the same data should behave the same to have similar conditions between installations, therefore the configured charset/collation should be applied when creating projection schemas.

This includes some additional optimizations for column definitions limiting the possible contents based on the currently existing limitations in the code.

Fixes: #4729

bwaidelich reviewed (commented):
Thanks a lot for investigating and taking care!

This looks great, I'm just wondering: With that we seem to leave the charset/collation up to the database default (I think?) – shouldn't we instead set it explicitly?

kitsunet commented:
I guess I should have mentioned that Flow does set a default for this, see

https://github.com/neos/flow-development-collection/blob/9.0/Neos.Flow/Configuration/Settings.Persistence.yaml#L17

I guess we could add a default collation as well to make it even more deterministic. Right now I assume the collaction is chosen by the database based on the charset.

kitsunet commented:
Note obviously that depends on the connection so as long as we use the default flow persistence connection this is set, if you declare your own connection for eventsourced stuff you would have to set this yourself.

bwaidelich commented:
> I guess I should have mentioned that Flow does set a default for this

Yes, I know. My point is, that with #4725 we rely on a collation so I don't think that this should be configurable!?

kitsunet commented:
Mmm, good point I guess, but it needs to be dynamic and match the charset? If we hardcode utf8mb4_unicode_ci and someone wants their DB in utf8mb3 charset that might be problematic. I think in #4725 we have to somehow derive a case sensitive and insensitive collation from the configured charset and use those ignoring whatever collation the db uses as default?

so basically we want to use {charset}_bin and {charset}_ci respectively.

Unlike a charset convert setting a collation for an operation is not adding overhead but simply being more specific in what exactly is supposed to happen.

bwaidelich commented:
> If we hardcode utf8mb4_unicode_ci and someone wants their DB in utf8mb3 charset

I wonder about the usecase for this and whether everything else still works if you change the charset/collation for the graph projection for example..

> so basically we want to use {charset}_bin and {charset}_ci respectively.

yes, right.. It might not be available for all charsets though?
Also I don't think that it's possible to retrieve the charset without additional queries (like `SELECT character_set_name FROM information_schema.COLUMNS WHERE ...`)

kitsunet commented:
> I wonder about the usecase for this and whether everything else still works if you change the charset/collation for the graph projection for example..

We have that example right now with the different defaults I observed and it seemed fine? I guess we could test something really exotic like latin or some other old ascii-ish format? I would definitely not want to hardcode this, also in view of potentially supporting other DBs via DBAL at some point _maybe_

We could _assume_ that the characterset is the one configured in the connection as default (the one we will use with this change) and base our decision on that, but yes it might still be the case that a fitting collation does not exist, but hardcoding seems really bad?

kitsunet commented:
Oh addendum we definitely should make sure the eventstore uses the same character set, it was the only that had code to use the configured one in flow already, so it was mb4 while the rest was mb3 (depending on system), at least not the other way around but that means you could store certain emoji in the eventstore but not project them afterwards.

bwaidelich commented:
btw: Some nice video on that topic: https://www.youtube.com/watch?v=ML24GRzXF9w (and a great channel for SQL topics in general!)

bwaidelich commented:
## Suggestion:

What if we allow charset and collations for CS & CI mode to be over-configured via the `options` that are already available to the factories, i.e.:

```yaml
Neos:
  ContentRepositoryRegistry:
    presets:
      'default':
        projections:
          'Neos.ContentRepository:ContentGraph':
            factoryObjectName: Neos\ContentGraph\DoctrineDbalAdapter\DoctrineDbalContentGraphProjectionFactory
            options:
              'database':
                charset: 'utf8mb4'
                collation_cs: 'utf8mb4_0900_as_cs'
                collation_ci: 'utf8mb4_0900_as_ci'
```

That way we could use sane defaults but still allow users to change the options _per_ Content Repository instance.

We _could_ make `charset`  default to whatever is configured in `Neos.Flow.persistence.backendOptions.charset` or `Neos.Flow.persistence.backendOptions.defaultTableOptions.charset` but that would create a soft dependency to Flow..
Instead I would make the default the same (`utf8mb4`) and use `utf8mb4_general_ci` and `utf8mb4_bin` for the default collations (at least, those seem to be available for MariaDB and MySQL)

kitsunet commented:
> We _could_ make `charset` default to whatever is configured in `Neos.Flow.persistence.backendOptions.charset` or `Neos.Flow.persistence.backendOptions.defaultTableOptions.charset` but that would create a soft dependency to Flow..
> Instead I would make the default the same (`utf8mb4`) and use `utf8mb4_general_ci` and `utf8mb4_bin` for the default collations (at least, those seem to be available for MariaDB and MySQL)

Sounds reasonable, but then we need to sort this out ourselfs again and can't rely on the doctrine api to create a schema configuration which feels a bit dirty. 

We should at least provide a little boilerplate helper then for projections so that we do not duplicate the sorting out of charsets for the schema everywhere, as I foresee people messing this up at some point.

kitsunet commented:
Little bonus conundrum, do we really want to allow different charsets between projections (and consequently the eventstore)? that seems potentially dangerous?

kitsunet commented:
We could actually add our own configuration to the default table options and retrieve them from there, there is no schema for it, seems dirty as well, but might be easier because it means you can always get it from the connection directly?

bwaidelich commented:
> Sounds reasonable, but then we need to sort this out ourselfs again and can't rely on the doctrine api to create a schema configuration which feels a bit dirty.

You mean that it would ignore the default charset/collation of the selected database?
That's true, but I don't know if that's a bad thing really – as long as it can be changed per CR

> We should at least provide a little boilerplate helper then for projections so that we do not
> duplicate the sorting out of charsets for the schema everywhere

+1

> Little bonus conundrum, do we really want to allow different charsets between projections
> (and consequently the eventstore)? that seems potentially dangerous?

I'm not yet deeply enough into the topic to provide a qualified response. But my gut feeling is that this ain't an issue because we don't rely on any ordering between read models.
I think it's rather dangerous to have some global default that is enforced for every projection

kitsunet commented:
Okay, I would adjust this to the plan, sounds good to me.

I think differetn charsets / collations are mostly fine, I guess you can footgun yourself by having different charsets and having unicode errors in some parts while it's fine in others (as in eg. urls vs. content or so), but that's probably okay. In the eventstore we would then apply a similar pattern I guess? Set it to UTF8MB4 though an comment a strong recommendation to leave it at that to avoid unicode errors, because if it's broken in the events there is no recovery.

kitsunet commented:
HIGHLY OPINIONATED, and needs a replay (or a few adjustment queries on the projections before updating the schema).

More info / ideas soon.

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -142,4 +175,21 @@ private function createRestrictionRelationTable(Schema $schema): void
+        return $schemaConfig;
+    }
+
+    private function getDefaultTextCollation(): string
+    {
+        return 'utf8mb4_unicode_520_ci';
+    }
```
Just wondering: Does this have to be an instance method? It could be static, but I would suggest to turn it into a private constant instead.
And don't worry for future changes: It is very local and only used in two places. If this happens to be dependent on the connection at some point, we can always change that lateron

kitsunet replied:
yeah exactly willdo

bwaidelich commented on Neos.ContentRepository.Core/Classes/SharedModel/Id/UuidFactory.php:
```diff
@@ -30,4 +30,16 @@ public static function create(): string
+    {
+        $uuid = self::create();
+
+        return str_replace('-', '', $uuid);
```
Is that the actual HEX representation of a UUID v4? And is it the responsibility of the UUID factory to create hex and bin representations? (probably not an issue, I'm just thinking out loud)

kitsunet replied:
seemed easier to do it here, I will check but I think this should be correct, the dashes are just added for readability.

bwaidelich reviewed (commented):
Wow, thanks for your work!

I have to admin, that I'm a little torn right now, because this changed from a bugfix to a complete overhaul with many changes that need explaining (at least to a db noob like me)..
E.g. (why) is binary a better representation for anchor and dimension space points? why is ascii better for node ids and types, why a collation for node ids but not for classification, ...

Also, renaming the root edge from `00000000-0000-0000-0000-000000000000` to  `00000000000000000000000000000000` might brake things?

Would you agree to split this up into two changes:
one with the charset/collation change, such that we can merge it immediately and rebase #4725 on top of
and one with the other, more opinionated, changes – maybe along side some more explanations for the reasoning that we can discuss and merge later if everyone agrees?

kitsunet commented:
Will split and annotate / explain. 

I might want to put some little helpers in the repository core to provide  schema defaults for specific "types" ala this is the schema for a nodeaggregateid kind of column as it was pretty nasty to get those in line between the differetn projections.

bwaidelich commented:
> Will split and annotate / explain.

Thank you! <3

> I might want to put some little helpers in the repository core

IMO we should be careful not to add unrelated utilities to the core or registry just because it's convenient. The core is deliberately independent from database specifics.
For now I would suggest to copy some code over the (not so many!) projections until we have a clearer understanding and then extract those to a new, separate, package (maybe it makes sense not to have a generic `neos/dbal-tools` but something along the line of `neos/contentrepository-dbaltools`)

kitsunet commented:
Some explanations already before splitting, because I would leave the _unproblematic_ charset/collation choices in here, that is:

some general considerations:

fixed length > variable length for performance
binary is slightly faster to compare than string
shorter length is better for index size
simpler charsets reduce storage and index size
simpler collations increase performance in like and sorts

binary datatype actually stores binary and will convert input to binary if needed, a varchar with binary charset will still store as a string type but doesn't keep any charset information so bytes as is, technically similar to the binary datatype but performance might still be a bit lower.

string-ish values that are pure identifiers, those need to compare exactly in all cases and we do not care about human/alphabetic ordering:

spacepointhash is a fixed size value (a hash) and can be represented binary without knowing anything about charsets

nodeaggregateid is a variable size string that is limited to ascii characters technically we could even make this charset binary (not binary type due to the char selection)

anchorpoints are identifiers (UUID and we should restrict it to that, upgrade PR would then the pure hex format), therefore fixed length and binary

classification, state and a few other strings that should actually be enums; as in they represent a fixed set of possible values. I am reducing the sizes to minimum and make them binary as we again do not care so much about the specific values but mostly comparing. better yet would be to declare an enum in sql or use ints for the possible values

strings with limited char selection that also needs to be sensibly sorted etc, example nodetypename, those can become ascii for storage size and performance

kitsunet commented:
> The core is deliberately independent from database specifics

the core requires dbal and has a couple of projections though?

bwaidelich commented:
> the core requires dbal and has a couple of projections though?

yes, you are right.. They were separate only in my dreams :)

kitsunet commented:
I do like that dream though!

kitsunet committed:
TASK: Deterministic projection schema with sensible defaults

Projections should be created with deterministic charset and collation and columns containing the same data should behave the same
to have similar conditions between installations, therefore the configured charset/collation should be applied when creating projection schemas.

This includes some additional optimizations for column definitions limiting the possible contents based on the currently existing limitations in the code.

Fixes: #4729

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -807,7 +807,7 @@ function (NodeRecord $node) use ($eventEnvelope) {
                         'nodeanchorpoint' => $nodeAnchorPoint?->value,
                         'destinationnodeaggregateid' => $reference->targetNodeAggregateId->value,
                         'properties' => $reference->properties
-                            ? \json_encode($reference->properties, JSON_THROW_ON_ERROR)
```
I can remove this one, overlooked it, the idea is to not create empty array in the DB `[]` but always objects `{}` seems cleaner, but it not necessary for aynthing I did below for now.

kitsunet committed:
Annotate basic schema columns

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -807,7 +807,7 @@ function (NodeRecord $node) use ($eventEnvelope) {
                         'destinationnodeaggregateid' => $reference->targetNodeAggregateId->value,
                         'properties' => $reference->properties
-                            ? \json_encode($reference->properties, JSON_THROW_ON_ERROR)
+                            ? \json_encode($reference->properties, JSON_THROW_ON_ERROR & JSON_FORCE_OBJECT)
```
I can remove this one, overlooked it, the idea is to not create empty array in the DB [] but always objects {} seems cleaner, but it not necessary for aynthing I did below for now.

bwaidelich replied:
Let's keep it, it makes a lot of sense!

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -29,27 +33,19 @@ public function buildSchema(): Schema
+        $table = DbalSchemaFactory::addColumnForDimensionSpacePoint($table, 'origindimensionspacepoint', false);
+        $table = DbalSchemaFactory::addColumnForDimensionSpacePointHash($table, 'origindimensionspacepointhash', false);
+        $table = DbalSchemaFactory::addColumnForNodeTypeName($table, 'nodetypename');
+        $table->addColumn('properties', Types::TEXT)
```
You removed that

> // TODO longtext?

comment. Was that on purpose? (Is there some `Types::LONGTEXT` at all?)

kitsunet replied:
There is not, but doctrine just creates a LONGTEXT if the length isn't limited as in this case, I first added a "this will result in a longtext as we do not restrict length" comment but found it silly because after this review no one will probably know why that comment is there in the first place.

bwaidelich replied:
Yes, totally agree, thanks for checking!

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:
```diff
@@ -29,27 +33,19 @@ public function buildSchema(): Schema
+        $table = DbalSchemaFactory::addColumnForDimensionSpacePointHash($table, 'origindimensionspacepointhash', false);
+        $table = DbalSchemaFactory::addColumnForNodeTypeName($table, 'nodetypename');
+        $table->addColumn('properties', Types::TEXT)
+            ->setNotnull(true)
```
Would it make sense to set a default value of `'{}'`?
At least that's what Planetscape recommends (https://youtu.be/cDzrY-QW4Ck?t=356)

kitsunet replied:
Ah good eye thanks! I removed it in my effort to separate the things, but it can go into this change!

bwaidelich commented on Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php:
```diff
@@ -0,0 +1,134 @@
+ *
+ * @internal Because we might need to check for platform later on and generally change the input and output format of functions within.
+ */
+final class DbalSchemaFactory
```
Factory is the new Service it seems ;)
Currently it's not really a factory, but I can't really think of a better name right now..

kitsunet replied:
Yep happy for any suggestions, it builds columns after all (or at least that was the idea when I named it factory).

ahaeslich replied:
As we have no other naming suggestion: do we want to keep it and get this change in?

mhsdesign replied:
Im fine with the name and its internal either way so we _can_ change it ;)

bwaidelich commented on Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php:
```diff
@@ -0,0 +1,134 @@
+     * The NodeAggregateId is limited to 64 ascii characters and therefore we should do the same in the database.
+     * @see NodeAggregateId
+     */
+    public static function addColumnForNodeAggregateId(Table $table, string $columnName, bool $notNull): Table
+    {
+        $table->addColumn($columnName, Types::STRING)
+                ->setLength(64)
+                ->setNotnull($notNull)
+                ->setCustomSchemaOption('charset', 'ascii')
+                ->setCustomSchemaOption('collation', 'ascii_general_ci');
+
+        return $table;
+    }
```
What about turning this into a simple:
```suggestion
    public static function columnForNodeAggregateId(string $name): Column
    {
        return (new Column($name, Type::getType(Types::STRING))
                ->setLength(64)
                ->setCustomSchemaOption('charset', 'ascii')
                ->setCustomSchemaOption('collation', 'ascii_general_ci');
    }
```
Unfortunately `Table->_addColumn()` is protected, but this could be used in the constructor..

kitsunet replied:
Yep I was also looking at _addColumn() and also would prefer that syntax, settled on what I have now though because then we start mixing between this syntax and addColumn (or write out everything in the constructor, which seemed a bit messy as well?

I can try it, but is it worth the effort right now?

bwaidelich replied:
Maybe it's not really worth it, but I fear that we'll have to add more parameters for upcoming edge cases..
So, just to see how this would feel like.
Instead of
```php
$table = $schema->createTable($this->tableNamePrefix . '_node');
$table = DbalSchemaFactory::addColumnForNodeAnchorPoint($table, 'relationanchorpoint');
$table = DbalSchemaFactory::addColumnForNodeAggregateId($table, 'nodeaggregateid', false);
// ...
```
it'd be
```php
$table = new Table($this->tableNamePrefix . '_node', [
    DbalSchemaFactory::columnForNodeAnchorPoint('relationanchorpoint'),
    DbalSchemaFactory::columnForNodeAggregateId('nodeaggregateid')->setNotnull(false),
    // ...
]);
```
If you agree to that, I could take care.

kitsunet committed:
Make phpcs happy

kitsunet commented:
I made this as a sort of memory to collect and decide on how to deal with UUIDs best in the future: https://github.com/neos/neos-development-collection/issues/4756

bwaidelich commented on Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php:
```diff
@@ -0,0 +1,138 @@
+     *
+     * @see ContentStreamId
+     */
+    public static function addColumnForContentStreamId(Table $table, string $columnName, bool $notNull): Table
+    {
+        $table->addColumn($columnName, Types::STRING)
+            ->setLength(36)
+            ->setNotnull($notNull)
```
If we decide against changing it as suggested above, I would at least suggest not to copy the double negative for nullable columns
```suggestion
    public static function addColumnForContentStreamId(Table $table, string $columnName, bool $nullable): Table
    {
        $table->addColumn($columnName, Types::STRING)
            ->setLength(36)
            ->setNotnull(!$nullable)
```

kitsunet replied:
wanted to keep it as close as possible to the original syntax so as to not mess things up, doing this you need to carefully inverse everything that was there before :D But probably a good call. 

Right now I have no other task so I can adjust the syntax as well, but I assime there would be more important stuff or either of us to spend time on? After all this is all internal and could be refactored later.

kitsunet committed:
Redo column generation methods for schemas

kitsunet commented:
messed one up, weirdly cr:setup did run on my maschine, so I pushed, well will fix later.

mhsdesign commented:
Just a note: i plan to further experiment with replacing the slow `$schemaManager->createSchema()` at some point see https://github.com/neos/neos-development-collection/issues/4762 

Does this somehow conflict with these changes? It seems not?

kitsunet commented:
I don't think so, should work just the same with just a few tables.

bwaidelich reviewed (approved):
This looks great (+1 by reading).
Thanks for taking care and bearing with me :)

kitsunet commented:
> This looks great (+1 by reading). Thanks for taking care and bearing with me :)

Only this way we make the code base excellent and maintainable so nothing to be be sorry about :)

mhsdesign commented:
+1 from me post merge as well ;) Works and looks great!

bwaidelich commented:
FYI: The binary column schemas were wrong and lead to `setUp()` calls to drop and re-create them. This will be fixed with #4847
